File: android/core/src/main/kotlin/com/clipvault/core/Normalize.kt
package com.clipvault.core

import java.security.MessageDigest
import java.text.Normalizer

/** NORM-1 (CONTRACTS §2). Must match desktop clipvault.core.normalize byte-for-byte. */
object Normalize {
    const val DEFAULT_MAX_CLIP_BYTES = 1_048_576
    const val REJECT_EMPTY = "empty"
    const val REJECT_TOO_LARGE = "too_large"

    fun normalize(text: String): String {
        var s = text.replace("\r\n", "\n").replace("\r", "\n")
        s = Normalizer.normalize(s, Normalizer.Form.NFC)
        return s.trimEnd()  // Kotlin trimEnd() strips Unicode whitespace, like Python rstrip()
    }

    fun rejectReason(normalized: String, maxBytes: Int = DEFAULT_MAX_CLIP_BYTES): String? {
        if (normalized.isEmpty()) return REJECT_EMPTY
        if (normalized.toByteArray(Charsets.UTF_8).size > maxBytes) return REJECT_TOO_LARGE
        return null
    }

    fun contentHash(normalized: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
            .digest(normalized.toByteArray(Charsets.UTF_8))
        return digest.joinToString("") { "%02x".format(it) }
    }
}

File: desktop/clipvault/core/normalize.py
"""NORM-1: content normalization and hashing (CONTRACTS §2).

Order is contractual: newline folding -> NFC -> strip end-of-string
whitespace. Inner-line trailing whitespace and all leading whitespace are
content and must be preserved.
"""

import hashlib
import unicodedata

REJECT_EMPTY = "empty"
REJECT_TOO_LARGE = "too_large"

DEFAULT_MAX_CLIP_BYTES = 1_048_576  # 1 MiB


def normalize(text: str) -> str:
    s = text.replace("\r\n", "\n").replace("\r", "\n")
    s = unicodedata.normalize("NFC", s)
    return s.rstrip()


def reject_reason(normalized: str, max_bytes: int = DEFAULT_MAX_CLIP_BYTES) -> str | None:
    """Return a rejection reason or None if the content is acceptable."""
    if normalized == "":
        return REJECT_EMPTY
    if len(normalized.encode("utf-8")) > max_bytes:
        return REJECT_TOO_LARGE
    return None


def content_hash(normalized: str) -> str:
    return hashlib.sha256(normalized.encode("utf-8")).hexdigest()
